### PR TITLE
implement BlockId

### PR DESCRIPTION
**Pull request by almogdepaz**

### CODE REVIEW

File: activation/activation.go [hunk 1]
```diff
@@ -559,3 +560,29 @@ func (b *Builder) GetPositioningAtx(epochId types.EpochId) (*types.ActivationTxH
 	}
 	return posAtx, nil
 }
```
Comment (almogdepaz) [inline activation/activation.go:562]: @noam moved here from types to avoid types being dependent of signing (circular dependency)

File: common/types/block.go [hunk 2]
```diff
@@ -142,18 +140,16 @@ func (b *BlockHeader) AddView(id BlockID) {
 	b.ViewEdges = append(b.ViewEdges, id)
 }
 
-func (b *Block) Compare(bl *Block) bool {
+func (b *Block) Compare(bl *Block) (bool, error) {
```
Comment (gavraz) [inline common/types/block.go:148]: Consider calling bytes.Compare. With compare you can check for lower, greater and equal.
You can look at the hare to see if it is useful there.
If you don't think there is a use for that then simply call the func Equal.

File: common/types/block.go [hunk 3]
```diff
@@ -173,11 +169,11 @@ func (l *Layer) Hash() Hash32 {
 	bids := l.blocks
 	keys := make([]BlockID, 0, len(bids))
 	for _, tortoiseBlock := range bids {
-		keys = append(keys, tortoiseBlock.ID())
+		keys = append(keys, tortoiseBlock.Id())
 	}
 	hash, err := CalcBlocksHash32(keys)
 	if err != nil {
-		log.Panic("failed to calculate layer's hash - layer Id %v", l.index)
+		panic(fmt.Sprintf("failed to calculate layer's hash - layer Id %v", l.index))
```
Comment (gavraz) [inline common/types/block.go]: Please call log.Panic so we have this log also in Kibana (format JSON, that is why we added .Panic in the first place).

File: common/types/block.go [hunk 4]
```diff
@@ -199,22 +195,18 @@ func NewExistingLayer(idx LayerID, blocks []*Block) *Layer {
 	return &l
 }
 
-func NewExistingBlock(id BlockID, layerIndex LayerID, data []byte) *Block {
+func NewExistingBlock(layerIndex LayerID, data []byte) *Block {
```
Comment (gavraz) [inline common/types/block.go:209]: Please make sure you have a UT that checks that blocks with the same data generate the same block id and block with different data generate different block id.

Comment (gavraz) [reply common/types/block.go:209]: (It should also make sure the id field is set, i.e. not the default value)

File: common/types/block.go [hunk 5]
```diff
@@ -223,3 +215,13 @@ func NewLayer(layerIndex LayerID) *Layer {
 		blocks: make([]*Block, 0, 10),
 	}
 }
+
```
Comment (gavraz) [inline common/types/block.go:229]: What happens if SortBlockIds and SortBlocks do not sort by the same order (i.e. ASC DESC)?
If it could have any significance I would bother to add a comment for both functions that describe this relation.

Comment (gavraz) [reply common/types/block.go:229]: Of course, if they have a relationship and you can enforce it it is even better. Maybe a 3rd function that compares to block ids. Then both functions can use it.

File: common/types/hashes.go [hunk 6]
```diff
@@ -198,3 +198,12 @@ func (h *Hash32) Scan(src interface{}) error {
 	copy(h[:], srcB)
 	return nil
 }
+
+func AllZero(s Hash32) bool {
```
Comment (gavraz) [inline common/types/hashes.go]: Why do you need this function?

File: events/events.go [hunk 7]
```diff
@@ -86,7 +86,7 @@ func (NewBlock) getChannel() channelId {
 }
 
 type ValidBlock struct {
-	Id    uint64
+	Id    string
```
Comment (gavraz) [inline events/events.go:89]: Why NewBlock and ValidBlock have a string id and not BlockID?

File: mesh/mesh.go [hunk 8]
```diff
@@ -278,63 +276,64 @@ func (m *Mesh) GetLatestView() []types.BlockID {
...
 	//new block add to orphans
-	m.handleOrphanBlocks(&blk.BlockHeader)
+	m.handleOrphanBlocks(blk)
 
 	//invalidate txs and atxs from pool
 	m.invalidateFromPools(&blk.MiniBlock)
 
-	events.Publish(events.NewBlock{Id: uint64(blk.ID()), Atx: blk.ATXID.ShortString(), Layer: uint64(blk.LayerIndex)})
-	m.With().Debug("added block", log.BlockId(uint64(blk.ID())))
+	events.Publish(events.NewBlock{Id: blk.Id().String(), Atx: blk.ATXID.ShortString(), Layer: uint64(blk.LayerIndex)})
```
Comment (gavraz) [inline mesh/mesh.go:336]: Maybe shortString? should it be readable?

File: common/types/block.go [hunk 9]
```diff
@@ -199,22 +195,18 @@ func NewExistingLayer(idx LayerID, blocks []*Block) *Layer {
...
 				LayerIndex: LayerID(layerIndex),
 				Data:       data},
 		}}
-	return &b
-}
 
-func RandBlockId() BlockID {
-	id := uuid.New()
-	return BlockID(binary.BigEndian.Uint64(id[:8]))
+	b.CalcAndSetId()
```
Comment (gavraz) [inline common/types/block.go:219]: What happens if the data changes after the block are created? will we re-calc the block id?
Make sure you have a UT for this scenario.

File: miner/block_builder.go [hunk 10]
```diff
@@ -221,15 +221,18 @@ func (t *BlockBuilder) createBlock(id types.LayerID, atxID types.AtxId, eligibil
...
-	t.Log.Event().Info(fmt.Sprintf("I've created a block in layer %v. id: %v, num of transactions: %v, votes: %d, viewEdges: %d atx %v, atxs:%v",
-		b.LayerIndex, b.ID(), len(b.TxIds), len(b.BlockVotes), len(b.ViewEdges), b.ATXID.ShortString(), len(b.AtxIds)))
-
 	blockBytes, err := types.InterfaceToBytes(b)
 	if err != nil {
 		return nil, err
 	}
 
-	return &types.Block{MiniBlock: b, Signature: t.Signer.Sign(blockBytes)}, nil
+	bl := &types.Block{MiniBlock: b, Signature: t.Signer.Sign(blockBytes)}
```
Comment (gavraz) [inline miner/block_builder.go:229]: Why not go with an additional "constructor" for blocks (just like NewExistingBlock) and it will call CalcAndSetId? It is safer.
This way it is the programmer's responsibility to call Calc after creating a block in such a way.

File: activation/activation.go [hunk 11]
```diff
@@ -559,3 +560,29 @@ func (b *Builder) GetPositioningAtx(epochId types.EpochId) (*types.ActivationTxH
...
+
+// ValidateSignedAtx extracts public key from message and verifies public key exists in IdStore, this is how we validate
```
Comment (antonlerner) [inline activation/activation.go]: change comment

Comment (almogdepaz) [reply activation/activation.go]: sure but this was the original comment you put in types...
(this was moved here from types to resolve circular dependencies)

File: sync/handler.go [hunk 12]
```diff
@@ -126,7 +126,7 @@ func newATxsRequestHandler(s *Syncer, logger log.Log) func(msg []byte) []byte {
 			if tx, err := s.atxpool.Get(t); err == nil {
 				atxs[t] = tx
 			} else {
-				logger.With().Error("error handling atx request message", log.AtxId(t.ShortString()), log.String("atx_ids", fmt.Sprintf("%x", atxids)))
+				logger.Error("error %v while handling atx request message ", err, log.String("atx_ids", fmt.Sprintf("%x", atxids)))
```
Comment (gavraz) [inline sync/handler.go]: Why did you remove "with" ? We need it for the JSON parsing...

File: sync/syncer.go [hunk 13]
```diff
@@ -511,7 +510,7 @@ func validateVotes(blk *types.Block, forBlockfunc ForBlockInView, depth int, lg
 	}
 
 	if err != nil {
-		lg.Error("validate votes failed", err)
+		lg.Error("validate votes failed for block %s err %s", blk.Id(), err)
```
Comment (gavraz) [inline sync/syncer.go]: Use "with"

File: sync/worker.go [hunk 14]
```diff
@@ -158,6 +158,7 @@ func NewFetchWorker(s WorkerInfra, count int, reqFactory BatchRequestFactory, id
 				idsStr := concatShortIds(remainingItems)
 				lg.Info("send fetch request to Peer: %v ids: %v", peer.String(), idsStr)
 				ch, _ := reqFactory(s, peer, remainingItems)
+				lg.Info(concatShortIds(remainingItems))
```
Comment (gavraz) [inline sync/worker.go]: Do you calculate it twice? concatShortIds is called twice

File: tortoise/algorithm.go [hunk 15]
```diff
@@ -29,7 +29,7 @@ func (alg *Algorithm) HandleLateBlock(b *types.Block) {
 	l := types.Layer{}
 	l.AddBlock(b)
 	alg.HandleIncomingLayer(&l)
-	log.Info("late block with layer id %v block id: %v ", b.Layer(), b.ID())
+	log.Info("late block with layer id %v block id: %v ", b.Layer(), b.Id())
```
Comment (gavraz) [inline tortoise/algorithm.go]: with

**Review by gavraz — CHANGES_REQUESTED** [review]

File: common/types/block.go (hunk 4 above)
Comment (almogdepaz) [reply common/types/block.go:209]: tested in the sync pkg

File: common/types/transaction.go [hunk 16]
```diff
@@ -110,32 +108,7 @@ func NewTxWithOrigin(nonce uint64, orig, rec Address, amount, gasLimit, fee uint
...
-		Signature:        [64]byte{},
-		origin:           &Address{},
-	}
-
-	copy(sst.Signature[:], signer.Sign(buf))
-	sst.origin.SetBytes(signer.PublicKey().Bytes())
-
-	return sst, nil
+type Signer interface {
 }
```
Comment (noamnelke) [inline common/types/transaction.go]: Leftovers?

Actually, I think that creating an interface like this is a better solution than to move `NewSignedTx` away from the other constructor.

The best solution, really, is to get rid of both constructors entirely, since they're only used in tests. At first they were needed because you couldn't set the private `origin` field, but now there's `SetOrigin()`. This requires some tedious work on test code, so if you're not up to it, create a tech-debt issue for this and assign it to me. But if you do that - I think it's better to leave the constructor here for now.

File: cmd/hare/hare.go [hunk 17]
```diff
@@ -80,8 +80,8 @@ func (app *HareApp) Cleanup() {
 func buildSet() []types.BlockID {
 	s := make([]types.BlockID, 200, 200)
 
-	for i := uint64(0); i < 200; i++ {
-		s = append(s, types.BlockID(i))
+	for i := types.LayerID(0); i < 200; i++ {
+		s = append(s, types.NewExistingBlock(i, []byte{}).Id())
```
Comment (noamnelke) [inline cmd/hare/hare.go]: Is each block really supposed to be in a different layer? @gavraz 

Maybe it doesn't matter, but I'd change it because it's confusing--seems intentional.

It's as easy to make `i` a `uint8` and use it as a byte in the data field.

File: common/types/block.go [hunk 18]
```diff
@@ -48,7 +52,6 @@ func (id NodeId) ShortString() string {
 }
 
 type BlockHeader struct {
-	Id               BlockID
```
Comment (noamnelke) [inline common/types/block.go:51]: I'm surprised we can get away with not having the ID in the block header... Won't this cause us to use the full block more often (bad)?

File: common/types/block.go [hunk 19]
```diff
@@ -93,7 +95,7 @@ func (t *Block) Data() interface{} {
 func (t *Block) Bytes() []byte {
 	bytes, err := InterfaceToBytes(t.MiniBlock)
 	if err != nil {
-		log.Panic("could not extract block bytes, %v", err)
+		panic(fmt.Sprintf("could not extract block bytes, %v", err))
```
Comment (noamnelke) [inline common/types/block.go]: Why like this?

File: common/types/block.go (hunk 2 above)
Comment (almogdepaz) [reply common/types/block.go:148]: we only use it for testing something in the sync so its kind of redundant

File: common/types/block.go (hunk 3 above)
Comment (almogdepaz) [reply common/types/block.go]: types cant be dependent on logs because there is a circular dependency

File: common/types/block.go (hunk 5 above)
Comment (almogdepaz) [reply common/types/block.go:229]: refactored

File: sync/worker.go (hunk 14 above)
Comment (almogdepaz) [reply sync/worker.go]: redundant print

File: hare/flows_test.go [hunk 20]
```diff
@@ -198,8 +198,8 @@ func (m *mockIdentityP) GetIdentity(edId string) (types.NodeId, error) {
 func buildSet() []types.BlockID {
 	s := make([]types.BlockID, 200, 200)
 
-	for i := uint64(0); i < 200; i++ {
-		s = append(s, types.BlockID(i))
+	for i := types.LayerID(0); i < 200; i++ {
+		s = append(s, types.NewExistingBlock(i, []byte{}).Id())
```
Comment (noamnelke) [inline hare/flows_test.go]: Again, this causes each new block to be in a different layer. It looks like it's related to the test logic, but it's not. Let's introduce variance through the data field.

File: miner/block_builder.go (hunk 10 above)
Comment (almogdepaz) [reply miner/block_builder.go:229]: the only use of it before serialising and broadcasting is for the log line below so theres no need

File: hare/preroundtracker_test.go [hunk 21]
```diff
@@ -12,16 +13,20 @@ const (
...
-var value3 = blockID{3}
-var value4 = blockID{4}
-var value5 = blockID{5}
-var value6 = blockID{6}
-var value7 = blockID{7}
-var value8 = blockID{8}
-var value9 = blockID{9}
-var value10 = blockID{10}
+func genBlockId(i int) types.BlockID {
+	return types.NewExistingBlock(types.LayerID(i), []byte{byte(i)}).Id()
```
Comment (noamnelke) [inline hare/preroundtracker_test.go]: Why different layer ID?

File: log/zap.go [hunk 22]
```diff
@@ -2,6 +2,7 @@ package log
 
 import (
 	"fmt"
+	"github.com/spacemeshos/go-spacemesh/common/types"
```
Comment (noamnelke) [inline log/zap.go]: Isn't this a problem? I assumed that's why we're not using any custom types in our logging methods.

File: common/types/block.go (hunk 9 above)
Comment (gavraz) [reply common/types/block.go:219]: Ok, please add doc to state that the block should not be changed.

File: sync/handler.go [hunk 23]
```diff
@@ -62,13 +62,13 @@ func newBlockRequestHandler(msh *mesh.Mesh, logger log.Log) func(msg []byte) []b
...
 		logger.Info("handle block request ids: %s", concatShortIds(blockids))
 		for _, bid := range blockids {
-			var id = util.BytesToUint64(bid.Bytes())
-			logger.Debug("handle block %v request", id)
-			blk, err := msh.GetBlock(types.BlockID(id))
+			logger.Info("handle block %s request", bid.ShortString())
+			blk, err := msh.GetBlock(types.BlockID(bid))
 			if err != nil {
-				logger.Error("Error handling block request message, with BlockID: %d and err: %v", id, err)
+				logger.Error("Error handling block request message, with BlockID: %s and err: %v", bid.ShortString(), err)
```
Comment (gavraz) [inline sync/handler.go]: log.with

**Review by gavraz — APPROVED** [review]

File: tortoise/ninja_tortoise.go [hunk 24]
```diff
@@ -303,16 +301,16 @@ func (ni *ninjaTortoise) updatePatternTally(newMinGood votingPattern, correction
 	}
 }
 
-func (ni *ninjaTortoise) getCorrEffCounter() (map[types.BlockID]vec, map[types.LayerID]int, func(b *types.BlockHeader)) {
+func (ni *ninjaTortoise) getCorrEffCounter() (map[types.BlockID]vec, map[types.LayerID]int, func(b *types.Block)) {
 	correctionMap := make(map[types.BlockID]vec)
 	effCountMap := make(map[types.LayerID]int)
-	foo := func(b *types.BlockHeader) {
+	foo := func(b *types.Block) {
```
Comment (noamnelke) [inline tortoise/ninja_tortoise.go:307]: I think this is a shame... If we'd put the id in the header we wouldn't need to work with full blocks here.

It's a little more involved to ensure we always have an id, since we need the full block to calculate it, but I think it's worth it. You can look at what we're doing in the ATXs.

**Review by noamnelke — CHANGES_REQUESTED** [review]

File: common/types/block.go (hunk 18 above)
Comment (almogdepaz) [reply common/types/block.go:51]: not really

File: common/types/transaction.go (hunk 16 above)
Comment (almogdepaz) [reply common/types/transaction.go]: someone should definitely refactor the transaction code, i started doing the interface but it didn't seem worth the effort just for some testing code
any way its out of the scope of this pr

File: log/zap.go (hunk 22 above)
Comment (almogdepaz) [reply log/zap.go]: fixed

File: tortoise/ninja_tortoise.go (hunk 24 above)
Comment (almogdepaz) [reply tortoise/ninja_tortoise.go:307]: @noamnelke  its cached leave the tortoise alone

File: common/types/transaction.go (hunk 16 above)
Comment (noamnelke) [reply common/types/transaction.go]: 👍🏻

But remove the leftover interface...

File: cmd/hare/hare.go (hunk 17 above)
Comment (gavraz) [reply cmd/hare/hare.go]: Intentional and only for testing but you are right it's confusing.
@almogdepaz fix for me

File: hare/flows_test.go (hunk 20 above)
Comment (gavraz) [reply hare/flows_test.go]: @almogdepaz same

File: hare/preroundtracker_test.go (hunk 21 above)
Comment (gavraz) [reply hare/preroundtracker_test.go]: @almogdepaz same
I use it to conveniently gen different block ids (which is all I care about).
I agree with Noam that it is confusing and easy to fix.

**Review by noamnelke — APPROVED** [review]
